### ISSUE TITLE
Numpy reshape and tf.reshape give different results in MNIST classification

### ISSUE THREAD

**Issue by Anjum48** [tensorflow/tensorflow#3396]
### Environment info

Operating System:

```
Linux anjum-Studio-XPS-1340 4.4.0-31-generic #50-Ubuntu SMP Wed Jul 13 00:07:12 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

Installed version of CUDA and cuDNN: 
Using CPU only version

If installed from binary pip package, provide:
1. Which pip package you installed: https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-0.9.0-cp35-cp35m-linux_x86_64.whl
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.: 
   `0.9.0`
### Steps to reproduce
1. Use the example from http://terrytangyuan.github.io/2016/06/09/scikit-flow-v09/#customized-model but edit to use the MNIST dataset instead.
2. Flatten the MNIST images from 8x8 to 1x64 using `numpy.reshape` before passing the features to the model. This gives around 95% accuracy
3. Flatten the MNIST images from 8x8 to 1x64 using `tf.reshape` within the model. This gives around 93% accuracy
### What have you tried?

The feature tensors within the two models are the same shape:
Numpy.reshape: `Tensor("input:0", shape=(?, 64), dtype=float32)`
tf.reshape: `Tensor("Reshape:0", shape=(?, 64), dtype=float32)`
### Logs or other output that would be helpful

Gist recreating the issue is here: https://gist.github.com/Anjum48/ca0a1f597a70b6b50b37fed2032a91d5

**Comment by concretevitamin** [comment]
Did you try isolating the issue, by looking at the two reshaped `features` and compare differences?  That might uncover bugs/issues better than comparing the end-to-end accuracy.

(I expect them to be equal.  And the differences in accuracy are just...machine learning? =D Did you train the `tf.reshape` version to full?)

**Comment by Anjum48** [comment]
Thanks for the reply! What's the best way of comparing the differences between the reshaped features? I'm afraid my tensor debugging doesn't extend beyond print (I posted the output above).

I initially thought it was ML accuracy, but I would have thought that 2% was pretty significant using MNIST? It also seems to be giving the same result each time which makes me think something else is up. Is there an easy way to fix the random seed between models?

**Comment by concretevitamin** [comment]
`tf.set_random_seed(1618)` or your favorite number should do it.

To compare, can't you `print(sess.run(tf_features))` and `print(np_features)?`

**Comment by Anjum48** [comment]
Setting the random seed as above still gives the same results - so it's not a stochastic issue.

I'm using skflow so i'm not sure how to get `print(sess.run(tf_features))` to work

**Comment by concretevitamin** [comment]
Maybe @terrytangyuan can take a look?  Marking as community support.

**Comment by terrytangyuan** [comment]
@Anjum48 Try printing features in your custom model definition. It's hard to dig into this without isolating the problem since there could be a lot underlying variations that lead to this inconsistency.

**Comment by Anjum48** [comment]
Hi @terrytangyuan, here's where I've put my prints

```
def my_model(features, target):
    tf.set_random_seed(1618)
    target = tf.one_hot(target, 10, 1, 0)  # Create 10x1 one-hot vectors
    print(features)
    features = layers.stack(features, layers.fully_connected, [10, 20, 10])
    prediction, loss = (tf.contrib.learn.models.logistic_regression_zero_init(features, target))
    train_op = tf.contrib.layers.optimize_loss(
        loss, tf.contrib.framework.get_global_step(), optimizer='Adagrad',
        learning_rate=0.1)
    return {'class': tf.argmax(prediction, 1), 'prob': prediction}, loss, train_op


def my_model2(features, target):
    tf.set_random_seed(1618)
    target = tf.one_hot(target, 10, 1, 0)  # Create 10x1 one-hot vectors
    features = tf.reshape(features, [-1, 8*8])  # Flatten 8x8 MNIST digits
    print(features)
    features = layers.stack(features, layers.fully_connected, [10, 20, 10])
    prediction, loss = (tf.contrib.learn.models.logistic_regression_zero_init(features, target))
    train_op = tf.contrib.layers.optimize_loss(
        loss, tf.contrib.framework.get_global_step(), optimizer='Adagrad',
        learning_rate=0.1)
    return {'class': tf.argmax(prediction, 1), 'prob': prediction}, loss, train_op
```

And here's the output (i'm using yesterday's build):

```
/usr/bin/python3.5 "/home/anjum/Python/numpy vs tf reshape.py"
(1149, 8, 8) (288, 8, 8) (360, 8, 8)
WARNING:tensorflow:Using temporary folder as model directory: /tmp/tmp68y2znql
WARNING:tensorflow:Using temporary folder as model directory: /tmp/tmpc8uirsil
WARNING:tensorflow:Setting feature info to TensorSignature(dtype=tf.float32, shape=TensorShape([Dimension(None), Dimension(64)]), is_sparse=False)
WARNING:tensorflow:Setting targets info to TensorSignature(dtype=tf.int64, shape=TensorShape([Dimension(None)]), is_sparse=False)
Tensor("input:0", shape=(?, 64), dtype=float32)
WARNING:tensorflow:Setting feature info to TensorSignature(dtype=tf.float32, shape=TensorShape([Dimension(None), Dimension(8), Dimension(8)]), is_sparse=False)
WARNING:tensorflow:Setting targets info to TensorSignature(dtype=tf.int64, shape=TensorShape([Dimension(None)]), is_sparse=False)
Tensor("Reshape:0", shape=(?, 64), dtype=float32)
Tensor("input:0", shape=(?, 64), dtype=float32)
Tensor("Reshape:0", shape=(?, 64), dtype=float32)
             precision    recall  f1-score   support

          0       0.97      0.91      0.94        33
          1       0.97      1.00      0.98        28
          2       0.94      0.97      0.96        33
          3       1.00      0.91      0.95        34
          4       0.96      0.98      0.97        46
          5       0.94      0.96      0.95        47
          6       1.00      0.97      0.99        35
          7       0.92      0.97      0.94        34
          8       0.97      0.93      0.95        30
          9       0.93      0.95      0.94        40

avg / total       0.96      0.96      0.96       360

             precision    recall  f1-score   support

          0       0.97      0.94      0.95        33
          1       0.96      0.96      0.96        28
          2       0.80      0.85      0.82        33
          3       0.93      0.82      0.87        34
          4       0.96      1.00      0.98        46
          5       0.94      0.94      0.94        47
          6       0.97      0.97      0.97        35
          7       1.00      0.97      0.99        34
          8       0.87      0.90      0.89        30
          9       0.88      0.90      0.89        40

avg / total       0.93      0.93      0.93       360


Process finished with exit code 0
```

**Comment by terrytangyuan** [comment]
I don't see you using numpy to flatten though.

**Comment by Anjum48** [comment]
Sorry - check the Gist. It's flattened before it's passed to the fit of the first model

**Comment by terrytangyuan** [comment]
I see. I think we need some results using two different reshape functions in order to see what the problem is, without getting into estimators first.

**Comment by vincentvanhoucke** [comment]
Closing this issue as stale. Feel free to reopen if there is any concrete action to be taken.
